debugd/internal/debugd/server/server.go

```diff
@@ -76,46 +83,51 @@ func (s *debugdServer) GetInfo(ctx context.Context, req *pb.GetInfoRequest) (*pb
 	return &pb.GetInfoResponse{Info: info}, nil
 }
 
-// UploadBootstrapper receives a bootstrapper binary in a stream of chunks and writes to a file.
-func (s *debugdServer) UploadBootstrapper(stream pb.Debugd_UploadBootstrapperServer) error {
-	startAction := deploy.ServiceManagerRequest{
-		Unit:   debugd.BootstrapperSystemdUnitName,
-		Action: deploy.Start,
-	}
-	var responseStatus pb.UploadBootstrapperStatus
-	defer func() {
-		if err := s.serviceManager.SystemdAction(stream.Context(), startAction); err != nil {
-			s.log.With(zap.Error(err)).Errorf("Starting uploaded bootstrapper failed")
-			if responseStatus == pb.UploadBootstrapperStatus_UPLOAD_BOOTSTRAPPER_SUCCESS {
-				responseStatus = pb.UploadBootstrapperStatus_UPLOAD_BOOTSTRAPPER_START_FAILED
-			}
-		}
-		stream.SendAndClose(&pb.UploadBootstrapperResponse{
-			Status: responseStatus,
+// UploadFiles receives a stream of files (each consisting of a header and a stream of chunks) and writes them to the filesystem.
+func (s *debugdServer) UploadFiles(stream pb.Debugd_UploadFilesServer) error {
+	s.log.Infof("Received UploadFiles request")
+	err := s.transfer.RecvFiles(stream)
+	switch err {
+	case nil:
+		s.log.Infof("Uploading files succeeded")
+	case filetransfer.ErrReceiveRunning:
+		s.log.Warnf("Upload already in progress")
+		return stream.SendAndClose(&pb.UploadFilesResponse{
+			Status: pb.UploadFilesStatus_UPLOAD_FILES_ALREADY_STARTED,
 		})
-	}()
-	s.log.Infof("Starting bootstrapper upload")
-	if err := s.streamer.WriteStream(debugd.BootstrapperDeployFilename, stream, true); err != nil {
-		if errors.Is(err, fs.ErrExist) {
-			// bootstrapper was already uploaded
-			s.log.Warnf("Bootstrapper already uploaded")
-			responseStatus = pb.UploadBootstrapperStatus_UPLOAD_BOOTSTRAPPER_FILE_EXISTS
-			return nil
+	case filetransfer.ErrReceiveFinished:
+		s.log.Warnf("Upload already finished")
+		return stream.SendAndClose(&pb.UploadFilesResponse{
+			Status: pb.UploadFilesStatus_UPLOAD_FILES_ALREADY_FINISHED,
+		})
+	default:
+		s.log.With(zap.Error(err)).Errorf("Uploading files failed")
+		return stream.SendAndClose(&pb.UploadFilesResponse{
+			Status: pb.UploadFilesStatus_UPLOAD_FILES_UPLOAD_FAILED,
+		})
+	}
+
+	files := s.transfer.GetFiles()
+	for _, file := range files {
+		if file.OverrideServiceUnit == "" {
+			continue
 		}
-		s.log.With(zap.Error(err)).Errorf("Uploading bootstrapper failed")
-		responseStatus = pb.UploadBootstrapperStatus_UPLOAD_BOOTSTRAPPER_UPLOAD_FAILED
-		return fmt.Errorf("uploading bootstrapper: %w", err)
+		// continue on error to allow other units to be overridden
+		_ = s.serviceManager.OverrideServiceUnitExecStart(stream.Context(), file.OverrideServiceUnit, file.TargetPath)
```

Comment:
Is this a use-case for `multierr` ?

Reply:
I think we would have to upgrade our go.mod to require go 1.20: https://github.com/golang/go/issues/53435
Opinions? @katexochen

Reply:
Also: go 1.20 is not yet released..

Reply:
We do have an imported `multierr` package ;)